### PR TITLE
SNOW v2

### PR DESCRIPTION
**Pull request by bakatzir**
- [x] Deleted deprecated commands
- [x] moved to py3
- [x] using client
- [x] Added the *custom_fields* argument to the ***servicenow-get-ticket*** command
- [x] UT and mocked responses

## Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/22328

## Screenshots
in issue

## Does it break backward compatibility?
   - [x] No

### CODE REVIEW

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 1]
```diff
@@ -0,0 +1,1692 @@
...
+DEFAULT_RECORD_FIELDS = {
+    'sys_id': 'ID',
+    'sys_updated_by': 'UpdatedBy',
+    'sys_updated_on': 'UpdatedAt',
+    'sys_created_by': 'CreatedBy',
+    'sys_created_on': 'CreatedAt'
+}
+
+
+def create_ticket_context(data: dict, ticket_type: str) -> dict:
```
Comment (liorblob) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:120]: Maybe we want a new design for this? consult with @Arsenikr

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 2]
```diff
@@ -0,0 +1,1692 @@
...
+    if not isinstance(data, list):
+        return create_ticket_context(data, ticket_type)
+
+    tickets = []
+    for d in data:
+        tickets.append(create_ticket_context(d, ticket_type))
+    return tickets
+
+
+def get_ticket_human_readable(tickets, ticket_type: str):
```
Comment (liorblob) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:160]: New design?

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 3]
```diff
@@ -0,0 +1,1692 @@
...
+        for f in arr_fields:
+            field = f.split('=')
+            if len(field) > 1:
+                dic_fields[field[0]] = field[1]
+
+    return dic_fields
+
+
+# Converts unicode elements of obj (incl. dictionary and list) to string recursively
+def unicode_to_str_recur(obj):
```
Comment (liorblob) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:274]: I don't think we need this in py3

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 4]
```diff
@@ -0,0 +1,1542 @@
...
+        '2': '2 - Medium',
+        '3': '3 - Low'
+    }
+
+    result = []
+    for ticket in tickets:
+
+        hr = {
+            'Number': ticket.get('number'),
+            'System ID': ticket['sys_id'],
```
Comment (yuvalbenshalom) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:170]: use get

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 5]
```diff
@@ -0,0 +1,1542 @@
...
+        if 'impact' in ticket:
+            hr['Impact'] = ticket_severity.get(ticket['impact'], ticket['impact'])
+        if 'urgency' in ticket:
+            hr['Urgency'] = ticket_severity.get(ticket['urgency'], ticket['urgency'])
+        if 'severity' in ticket:
+            hr['Severity'] = ticket_severity.get(ticket['severity'], ticket['severity'])
+        if 'priority' in ticket:
+            hr['Priority'] = TICKET_PRIORITY.get(ticket['priority'], ticket['priority'])
+        if 'state' in ticket:
+            mapped_state = ticket['state']
```
Comment (yuvalbenshalom) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:198]: use get

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 6]
```diff
@@ -0,0 +1,1542 @@
...
+            'Additional Comments': ticket.get('comments')
+        }
+
+        # Try to map the fields
+        if 'impact' in ticket:
+            hr['Impact'] = ticket_severity.get(ticket['impact'], ticket['impact'])
+        if 'urgency' in ticket:
+            hr['Urgency'] = ticket_severity.get(ticket['urgency'], ticket['urgency'])
+        if 'severity' in ticket:
+            hr['Severity'] = ticket_severity.get(ticket['severity'], ticket['severity'])
```
Comment (yuvalbenshalom) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:194]: is ticket severity is the same 3 values for impact and urgency by default?
and also, it is probably safer to use get one time for every value and use it.
like that:
```
impact = ticket.get('impact', '')
hr['Impact'] = ticket_severity.get(impact, impact)
```

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 7]
```diff
@@ -0,0 +1,1542 @@
...
+    }
+
+    # These fields refer to records in the database, the value is their system ID.
+    if 'closed_by' in data:
+        context['ResolvedBy'] = data['closed_by']['value'] if 'value' in data['closed_by'] else ''
+    if 'opened_by' in data:
+        context['OpenedBy'] = data['opened_by']['value'] if 'value' in data['opened_by'] else ''
+        context['Creator'] = data['opened_by']['value'] if 'value' in data['opened_by'] else ''
+    if 'assigned_to' in data:
+        context['Assignee'] = data['assigned_to']['value'] if 'value' in data['assigned_to'] else ''
```
Comment (yuvalbenshalom) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:117]: we can use `data['closed_by'].get('value, '')` and drop al the last if statements.

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 8]
```diff
@@ -0,0 +1,1542 @@
...
+
+    # Try to map fields
+    if 'priority' in data:
+        context['Priority'] = TICKET_PRIORITY.get(data['priority'], data['priority'])
```
Comment (yuvalbenshalom) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:121]: get priority before with get()

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 9]
```diff
@@ -0,0 +1,1542 @@
...
+    if 'state' in data:
+        context['State'] = data['state']
```
Comment (yuvalbenshalom) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:123]: get

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 10]
```diff
@@ -0,0 +1,1542 @@
...
+        if 'urgency' in ticket:
+            hr['Urgency'] = ticket_severity.get(ticket['urgency'], ticket['urgency'])
+        if 'severity' in ticket:
+            hr['Severity'] = ticket_severity.get(ticket['severity'], ticket['severity'])
+        if 'priority' in ticket:
+            hr['Priority'] = TICKET_PRIORITY.get(ticket['priority'], ticket['priority'])
+        if 'state' in ticket:
+            mapped_state = ticket['state']
+            if ticket_type in TICKET_STATES:
+                mapped_state = TICKET_STATES[ticket_type].get(ticket['state'], mapped_state)
```
Comment (yuvalbenshalom) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:200]: .get('state')

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 11]
```diff
@@ -0,0 +1,1542 @@
...
+
+        url = f'{self._base_url}{path}'
+        if not headers:
+            headers = {
+                'Accept': 'application/json',
+                'Content-Type': 'application/json'
+            }
+        if file:
+            # Not supported in v2
+            url = url.replace('v2', 'v1')
```
Comment (yuvalbenshalom) [inline Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:348]: what dose this do?

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py (hunk 6 above)
Comment (bakatzir) [reply Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:194]: Yes to the first question.

File: Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py (hunk 11 above)
Comment (bakatzir) [reply Packs/ServiceNowv2/Integrations/ServiceNowv2/ServiceNowv2.py:348]: if you want to upload a file to an incident/ticket you can not do it using the v2 of the API. so we're changing it.
https://docs.servicenow.com/bundle/geneva-servicenow-platform/page/integrate/inbound_rest/reference/r_AttachmentAPI-POSTmultipart.html
@liorblob correct me if needed.

**Comment by bakatzir** [comment]
@kirbles19 we have created a v2 of the integration, with many enhancements to follow.
can you review the yml so I can generate a detailed readme.
also, @liorblob please comment here/dm me anything that I should add to the docs.

**Comment by bakatzir** [comment]
SSL error:
Screen Shot 2020-04-10 at 1 06 15

creds error:
Screen Shot 2020-04-10 at 1 06 32

@yuvalbenshalom Not sure I understood regarding ticket actions.
let us do a small intervention with @liorblob about it on Sunday.

File: Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_description.md [hunk 12]
```diff
@@ -0,0 +1 @@
+
```
Comment (anara123) [inline Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_description.md:1]: We have nothing in the detailed description?

Comment (bakatzir) [reply Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_description.md:1]: Not from the previous integration, but i'll add something. the access to tables was a reoccurring issue on slack conversations.

**Comment by bakatzir** [comment]
@kirbles19 merging without your review as agreed. will put as assignee on the next PR.

**Comment by bakatzir** [comment]
reference: https://github.com/demisto/etc/issues/23708

**Comment by bakatzir** [comment]
fixes: https://github.com/demisto/etc/issues/17863

**Pull request merged**